### PR TITLE
media-sound/mumble: Update/Fix live ebuild

### PR DESCRIPTION
**Pull request by DarkDefender**
Upstream now has dropped qmake and uses only cmake.
Updated the ebuild to use cmake instead of qmake.

Also added tests.

Some minor fixes needed to be done to the build system.
I've submitted two pull requests upstream:
https://github.com/mumble-voip/mumble/pull/4466
https://github.com/mumble-voip/mumble/pull/4468

I'm submitting this now so I can get feedback on the ebuild side of things.

### CODE REVIEW

**Comment by DarkDefender** [comment]
I'm a bit unsure about the multilib nature of this ebuild.

Perhaps it was needed in the past, but now on x86_64, the default cmake build does build a x86 (32bit) overlay library so it can be used in 32bit games.

**Comment by DarkDefender** [comment]
My patches have now been merged upstream so this should be good to go unless there is some changes need to the ebuild.

**Comment by DarkDefender** [comment]
@thesamesam how do I poke the maintainers that are not on github?
It seems like I need to get in touch with `polynomial-c`

Do I file a ticket on the gentoo bug tracker?

**Comment by Ionic** [comment]
NAK from me.

This PR adds a subtle difference: actual support for multilib, i.e., multiple arch builds. This was disabled in the previous version for good reasons: `qmake` can't easily handle multilib stuff (boo), but more importantly, the Qt libraries are not multilib-compatible for some stupid reasons, so this ebuild will fail on any system that enables non-native arches by default. It probably would be good to switch back to `src_configure` again if that makes it fake-multilib for the native arch only.

Also, pinging @Polynomial-C.

**Comment by Ionic** [comment]
Additionally, please make sure to add

`Closes: https://bugs.gentoo.org/741716`

to the commit and read the [Gentoo Certificate of Origin](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin) and sign-off all your commits (`--signoff` option to commit).

**Comment by DarkDefender** [comment]
I've removed the multi lib things and cleaned up the ebuild.

If the `overlay` cmake flag is on, mumble will automatically build a 32bit library if the arch is amd64/x86_64.
So in this case we do not need to do any multi lib magic.

The building of this 32bit library can be turned off with `overlay-xcompile`, but I don't know how useful it will be to expose this in the ebuild. So I now let the build scripts handle this flag.

I have added the requested lines to the commit message.

I've also created a pull request upstream to fix the install rules for the overlay scripts and libs.
As this is a live ebuild I didn't include this patch here as it will probably be merged soon:
https://github.com/mumble-voip/mumble/pull/4495

**Comment by DarkDefender** [comment]
My latest upstream fix is now merged, so this should be good to go.

**Comment by DarkDefender** [comment]
@Ionic Anything preventing this from being merged now?

**Comment by Ionic** [comment]
Mh, not sure. The whole multilib business with the overlay library seems a bit weird, but could be okay, I guess. I'm not sure if it's self-contained or needing other non-native-arch libraries for instance, and given that it's turned on by default, I wonder if it behaves correctly in both cases - foreign arches being enabled and disabled.

I'm not a Gentoo dev, though, merely went through this because of personal interest and reviewed it a bit.

I guess you'll have to get @Polynomial-C's attention for this PR to move on.

**Comment by Polynomial-C** [comment]
This PR is on my todo list but I couldn't give it the attention it deserves yet.
AFAIK, the multilib stuff is important if you have "old" 32bit games that need the libmumble.so for the mumble overlay to work. I hope we can still keep that avaiable because otherwise this could be seen as a regression.

**Comment by DarkDefender** [comment]
The 32bit libs are now automatically built on x86_64 build targets.

So this updated ebuild should still provide support for older 32bit games on a multi lib system.

**Comment by kripton** [comment]
@DarkDefender: Thanks for doing the work on media-sound/mumble! I've created #17991 that does the same to media-sound/murmur. I'll also keep watching this PR to check all remarks by @Polynomial-C  and apply them to my PR as well (if they are valid for murmur)

File: media-sound/mumble/mumble-9999.ebuild
Comment (kripton) [inline media-sound/mumble/mumble-9999.ebuild:75]: Any reasons why you set this to `OFF`? Does creating the translations need any additional dependencies?

Comment (DarkDefender) [reply media-sound/mumble/mumble-9999.ebuild:75]: It is simply mirroring the old settings in the ebuild.
The old flag was "no-embed-qt-translations".

I don't remember if I tried turning it on or not.

But I guess we can have a discussion if we should make this configurable or not if needed.

Comment (kripton) [reply media-sound/mumble/mumble-9999.ebuild:75]: Ah, okay. I've just realised that with that setting, mumble is english-only. Setting it to `ON` makes it translated (or English with `LC_ALL=C`). However, since no additional translation-files are installed, I assume they are integrated into the binary.
I will re-check with mumble-1.3.3 tomorrow to see how that behaves.
However, it might be that integrating the translation files might require `dev-qt/linguist-tools` as dependency.

Comment (kripton) [reply media-sound/mumble/mumble-9999.ebuild:75]: Okay, I've re-checked mumble-1.3.3: While it is true that the ebuild requests the translations not to be embedded, this doesn't seem to be honored. `lrelease` from `dev-qt/linguist-tools` is called during the build for each language and the resulting binary displays the texts translated. Since the emerge doesn't install any translation files, I assume that the texts are inside the binary (tried `strings` command but didn't find anything. Did not want to investigate how Qt's resource system (qrc) works in detail).
With the CMake-based build system, the configuration option is actually taken into account.
I'd prefer mumble to simply integrate the translations. I wouldn't know how to force them to separate files.

**Comment by kripton** [comment]
And in my murmur-9999-PR (#17991) I got this review that is also valid for your ebuild: https://github.com/gentoo/gentoo/pull/17991#pullrequestreview-515015754

**Comment by DarkDefender** [comment]
@kripton are you sure? When I compile mumble I can see that my flags some make.conf is honored

**Comment by DarkDefender** [comment]
Hmm, I guess we don't want to pull in any extra compile flags, so that is why.

**Comment by kripton** [comment]
> @kripton are you sure? When I compile mumble I can see that my flags from make.conf is honored

Honestly, I simply don't know ;)
Other ebuilds in the tree using the cmake eclass have the debug flag in IUSE but don't have any reference to that flag in the ebuild: https://github.com/gentoo/gentoo/blob/64fe952caa63f4916a5e2f704ebd5306ff00198f/app-admin/keepassxc/keepassxc-2.6.2.ebuild#L27

And the cmake eclass docs mention to not change CMAKE_BUILD_TYPE:
```You usually do *NOT* want nor need to set it as it pulls CMake default build-type specific compiler flags overriding make.conf``` (https://devmanual.gentoo.org/eclass-reference/cmake.eclass/index.html)

**Comment by Polynomial-C** [comment]
So I finally found some time to give this PR some testing. Unfortunately I don't get any 32bit overlay libs on my amd64-multilib system.

**Comment by DarkDefender** [comment]
@Polynomial-C Is does install them on my system. I didn't move them from the default install location as I guess the lib only work with the 64bit client. (But works with 32bit games).

`equery f mumble`:
```
/usr/lib64/mumble
/usr/lib64/mumble/libcelt0.so -> libcelt0.so.0.7.0
/usr/lib64/mumble/libcelt0.so.0.7.0
/usr/lib64/mumble/libmumbleoverlay.x86.so -> libmumbleoverlay.x86.so.1.4.0
/usr/lib64/mumble/libmumbleoverlay.x86.so.1.4.0
/usr/lib64/mumble/libmumbleoverlay.x86_64.so -> libmumbleoverlay.x86_64.so.1.4.0
/usr/lib64/mumble/libmumbleoverlay.x86_64.so.1.4.0
```

File: media-sound/mumble/mumble-9999.ebuild
Comment (Polynomial-C) [inline media-sound/mumble/mumble-9999.ebuild:87]: Can you please adjust the quoting to look like this?
```
-Dfeature1="ON"
-Dfeature2="OFF"
-Dfeature3="$(usex feature3)"
```

File: media-sound/mumble/mumble-9999.ebuild
Comment (Polynomial-C) [reply media-sound/mumble/mumble-9999.ebuild:75]: According to the CMakeLists.txt file `-Dtranslation` triggers other tranlations than english. So please introduce a `nls` USE flag and change the option to `-Dtranslation="$(usex nls)"`

File: media-sound/mumble/mumble-9999.ebuild
Comment (Polynomial-C) [inline media-sound/mumble/mumble-9999.ebuild:53]: Move this into the non-conditional block above. Keep it alphabetically sorted please.

File: media-sound/mumble/mumble-9999.ebuild
Comment (Polynomial-C) [reply media-sound/mumble/mumble-9999.ebuild:87]: Oh and please sort the list alphabetically

File: media-sound/mumble/mumble-9999.ebuild
Comment (Polynomial-C) [inline media-sound/mumble/mumble-9999.ebuild:6]: Why did you replace `xdg.eclass` with `xdg-utils.eclass`?

**Review by Polynomial-C — CHANGES_REQUESTED** [review]

**Comment by Polynomial-C** [comment]
> @Polynomial-C Is does install them on my system. I didn't move them from the default install location as I guess the lib only works with the 64bit client. (But works with 32bit games).
> 
> `equery f mumble`:
> 
> ```
> /usr/lib64/mumble
> /usr/lib64/mumble/libcelt0.so -> libcelt0.so.0.7.0
> /usr/lib64/mumble/libcelt0.so.0.7.0
> /usr/lib64/mumble/libmumbleoverlay.x86.so -> libmumbleoverlay.x86.so.1.4.0
> /usr/lib64/mumble/libmumbleoverlay.x86.so.1.4.0
> /usr/lib64/mumble/libmumbleoverlay.x86_64.so -> libmumbleoverlay.x86_64.so.1.4.0
> /usr/lib64/mumble/libmumbleoverlay.x86_64.so.1.4.0
> ```

These libs need to go into the 32bit libdir (on 17.0 and older multilib profiles this is `/usr/lib32` and on 17.1 multilib profile this is `/usr/lib`)

**Comment by DarkDefender** [comment]
> > @Polynomial-C Is does install them on my system. I didn't move them from the default install location as I guess the lib only works with the 64bit client. (But works with 32bit games).
> > `equery f mumble`:
> > ```
> > /usr/lib64/mumble
> > /usr/lib64/mumble/libcelt0.so -> libcelt0.so.0.7.0
> > /usr/lib64/mumble/libcelt0.so.0.7.0
> > /usr/lib64/mumble/libmumbleoverlay.x86.so -> libmumbleoverlay.x86.so.1.4.0
> > /usr/lib64/mumble/libmumbleoverlay.x86.so.1.4.0
> > /usr/lib64/mumble/libmumbleoverlay.x86_64.so -> libmumbleoverlay.x86_64.so.1.4.0
> > /usr/lib64/mumble/libmumbleoverlay.x86_64.so.1.4.0
> > ```
> 
> These libs need to go into the 32bit libdir (on 17.0 and older multilib profiles this is `/usr/lib32` and on 17.1 multilib profile this is `/usr/lib`)

It's only one lib `libmumbleoverlay.x86.so`. Is there any automagical way to install into the right place (depending on the profile)?

File: media-sound/mumble/mumble-9999.ebuild
Comment (DarkDefender) [reply media-sound/mumble/mumble-9999.ebuild:6]: I don't really remember, but I think it was because I needed some functions that were not in the xdg class.
After the update I didn't need the xdg eclass

**Comment by Polynomial-C** [comment]
> Is there any automagical way to install into the right place (depending on the profile)?

I don't know any portage function that could help us here. You can get the correct libdir name for x86 by calling `$(get_abi_var LIBDIR x86)` Which either returns `lib32` for 17.0 and older amd64-multilib profiles and `lib` for 17.1 amd64-multilib profiles. Perhaps an additional cmake option that overrides the default install location for the 32bit overlay files could be useful here. Something like: `-D32bit_libdir="/usr/$(get_abi_var LIBDIR x86)"`
Of course, for pure 32bit x86 installations, all this is not necessary as the libdir is always `lib` there.

**Comment by DarkDefender** [comment]
@Polynomial-C I will hopefully update the lib32 stuff later today :crossed_fingers: 

Can you check if my update is good and that this is the only thing left that needs to be fixed?

File: media-sound/mumble/mumble-9999.ebuild
Comment (Polynomial-C) [reply media-sound/mumble/mumble-9999.ebuild:6]: `xdg.eclass` inhrits `xdg-utils.eclass` so I suppose you did it for some other reason. Anyway, please prefer `xdg.eclass`.

File: media-sound/mumble/mumble-9999.ebuild
Comment (Polynomial-C) [inline media-sound/mumble/mumble-9999.ebuild:87]: Still not alphabetically sorted

File: media-sound/mumble/mumble-9999.ebuild
Comment (Polynomial-C) [inline media-sound/mumble/mumble-9999.ebuild:95]: Please change that back when you use `xdg.eclass` again.

**Review by Polynomial-C — CHANGES_REQUESTED** [review]
See my other change request comments

File: media-sound/mumble/mumble-9999.ebuild
Comment (DarkDefender) [reply media-sound/mumble/mumble-9999.ebuild:95]: Ah, now I remember why.
The ebuild fails if I simply use `xdg` with:
```
FATAL: cmake_src_prepare has not been run
```

That is why I changed it.
So it seems like the xdg eclass messes up.

It also seems like there are quite a few other ebuilds that simply uses the utils class to not have to deal with this either.

Comment (Polynomial-C) [reply media-sound/mumble/mumble-9999.ebuild:95]: Just add
```
src_prepare() {
        # required because of xdg.eclass also providing src_prepare
        cmake_src_prepare
}
```
to the ebuild.

**Comment by DarkDefender** [comment]
There, now everything should have been taken care of :crossed_fingers:

File: media-sound/mumble/mumble-9999.ebuild
Comment (Polynomial-C) [inline media-sound/mumble/mumble-9999.ebuild:105]: Make these two variables `local`

File: media-sound/mumble/mumble-9999.ebuild
Comment (Polynomial-C) [inline media-sound/mumble/mumble-9999.ebuild:107]: Please use `${D}`, `${libdir_64}` and `${libdir_32}` (curly braces)

**Review by Polynomial-C — CHANGES_REQUESTED** [review]
I know I start to nitpick now but this is good ebuild writing style...

File: media-sound/mumble/mumble-9999.ebuild
Comment (Polynomial-C) [inline media-sound/mumble/mumble-9999.ebuild:106]: That line can be reduced by using dodir (sorry, didn't see it when writing my previous comment):
`dodir {$libdir_32}`
but no need to add this with another commit. I will change that when merging this PR.

**Comment by kripton** [comment]
@Polynomial-C : Can you please also review related #17991 ? I've watched your change-requests here and adapted my PR when necessary. But due to not having 32bit libraries, it's a bit simpler ;) I can also squash all commits into one if preferred

**Comment by Polynomial-C** [comment]
Thank you very much for your work and your patience with me.
FYI, I did some final adjustments in commit 3629693f926f2a47c4adc08e28987bb265a7aa4a

**Pull request closed without merging**
